File: Shared/RelaxKonOS.Protocol/ServerCenter/ServerReleaseArchiveVerifier.cs
using System.IO.Compression;
using System.Security.Cryptography;
using System.Text.Json;
using RelaxKonOS.Protocol.Common;

namespace RelaxKonOS.Protocol.ServerCenter;

/// <summary>客户端在 SFTP 上传前对整个离线发布 ZIP 进行验证的结果。</summary>
public sealed record ServerReleaseArchiveVerification(
    string? ProblemCode, ServerReleaseManifestDto? Manifest, string? ArchiveSha256)
{
    public bool Verified => ProblemCode is null && Manifest is not null;
}

/// <summary>
/// 发布签名绑定原始 manifest.json，清单绑定包内每个文件；ZIP 摘要只负责传输完整性。
/// 信任根由客户端内置或显式开发配置提供，包中携带的公钥永远不被自动信任。
/// </summary>
public static class ServerReleaseArchiveVerifier
{
    private const int MaximumManifestBytes = 1024 * 1024;
    private const int MaximumSignatureBytes = 16 * 1024;
    private const int MaximumEntries = 20000;
    private const long MaximumPayloadBytes = 8L * 1024 * 1024 * 1024;

    /// <summary>Verify the archive, then materialize only its signed files into a new directory.</summary>
    public static ServerReleaseArchiveVerification VerifyAndExtract(
        Stream archiveStream,
        string destination,
        ServerReleasePackageKind expectedKind,
        ServerRuntimeIdentifier expectedRuntime,
        IReadOnlyDictionary<string, string> trustedPublicKeys,
        string? expectedArchiveSha256 = null)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(destination);
        var result = Verify(archiveStream, expectedKind, expectedRuntime, trustedPublicKeys, expectedArchiveSha256);
        if (!result.Verified) return result;

        var target = Path.GetFullPath(destination);
        if (Directory.Exists(target) || File.Exists(target))
            throw new IOException("The extraction destination already exists.");
        var parent = Path.GetDirectoryName(target) ?? throw new IOException("The extraction destination has no parent.");
        Directory.CreateDirectory(parent);
        var temporary = Path.Combine(parent, ".release-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(temporary);
        try
        {
            archiveStream.Position = 0;
            using var archive = new ZipArchive(archiveStream, ZipArchiveMode.Read, leaveOpen: true);
            if (archive.Entries.Count > MaximumEntries)
                throw new InvalidDataException("The release contains too many entries.");
            var signedFiles = result.Manifest!.Files.ToDictionary(item => item.Path, StringComparer.Ordinal);
            foreach (var entry in archive.Entries)
            {
                if (entry.FullName.EndsWith("/", StringComparison.Ordinal)) continue;
                var relative = entry.FullName;
                if (!ServerReleaseValidation.IsSafeManifestPath(relative) ||
                    ((entry.ExternalAttributes >> 16) & 0xF000) == 0xA000 ||
                    relative is not ("manifest.json" or "manifest.json.sig") && !signedFiles.ContainsKey(relative))
                    throw new InvalidDataException("The archive changed during extraction.");
                var output = Path.Combine(temporary, relative.Replace('/', Path.DirectorySeparatorChar));
                Directory.CreateDirectory(Path.GetDirectoryName(output)!);
                using var input = entry.Open();
                using var file = new FileStream(output, FileMode.CreateNew, FileAccess.Write, FileShare.None);
                using var digest = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
                var buffer = new byte[81920];
                long length = 0;
                int count;
                while ((count = input.Read(buffer)) > 0)
                {
                    length = checked(length + count);
                    if (length > entry.Length) throw new InvalidDataException("The archive changed during extraction.");
                    file.Write(buffer, 0, count);
                    digest.AppendData(buffer, 0, count);
                }
                if (length != entry.Length) throw new InvalidDataException("The archive changed during extraction.");
                if (signedFiles.TryGetValue(relative, out var signed) &&
                    (length != signed.Length ||
                     !Convert.ToHexString(digest.GetHashAndReset()).Equals(signed.Sha256, StringComparison.OrdinalIgnoreCase)))
                    throw new InvalidDataException("A signed file changed during extraction.");
                if (OperatingSystem.IsLinux())
                {
                    var permissions = (UnixFileMode)((entry.ExternalAttributes >> 16) & 0x1FF);
                    File.SetUnixFileMode(output, permissions == 0
                        ? UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.GroupRead | UnixFileMode.OtherRead
                        : permissions);
                }
            }
            archiveStream.Position = 0;
            if (!HashStream(archiveStream).Equals(result.ArchiveSha256, StringComparison.Ordinal))
                throw new InvalidDataException("The archive changed during extraction.");
            Directory.Move(temporary, target);
            return result;
        }
        finally
        {
            if (Directory.Exists(temporary)) Directory.Delete(temporary, recursive: true);
        }
    }

    public static ServerReleaseArchiveVerification Verify(
        Stream archiveStream,
        ServerReleasePackageKind expectedKind,
        ServerRuntimeIdentifier expectedRuntime,
        IReadOnlyDictionary<string, string> trustedPublicKeys,
        string? expectedArchiveSha256 = null)
    {
        ArgumentNullException.ThrowIfNull(archiveStream);
        ArgumentNullException.ThrowIfNull(trustedPublicKeys);
        if (!archiveStream.CanRead || !archiveStream.CanSeek)
            throw new ArgumentException("A readable, seekable archive stream is required.", nameof(archiveStream));

        var originalPosition = archiveStream.Position;
        try
        {
            archiveStream.Position = 0;
            var archiveDigest = HashStream(archiveStream);
            if (expectedArchiveSha256 is not null &&
                (!ServerDeploymentInputRules.IsSha256(expectedArchiveSha256) ||
                 !string.Equals(archiveDigest, expectedArchiveSha256, StringComparison.OrdinalIgnoreCase)))
                return new(ServerDeploymentProblemCodes.PackageDigestMismatch, null, archiveDigest);

            archiveStream.Position = 0;
            using var archive = new ZipArchive(archiveStream, ZipArchiveMode.Read, leaveOpen: true);
            if (archive.Entries.Count > MaximumEntries)
                return new(ServerDeploymentProblemCodes.PackageLayoutUnsafe, null, archiveDigest);
            var entries = new Dictionary<string, ZipArchiveEntry>(StringComparer.Ordinal);
            foreach (var entry in archive.Entries)
            {
                if (entry.FullName.EndsWith("/", StringComparison.Ordinal))
                {
                    if (!ServerReleaseValidation.IsSafeManifestPath(entry.FullName.TrimEnd('/')))
                        return new(ServerDeploymentProblemCodes.PackageLayoutUnsafe, null, archiveDigest);
                    continue;
                }
                if (!ServerReleaseValidation.IsSafeManifestPath(entry.FullName) ||
                    !entries.TryAdd(entry.FullName, entry) ||
                    ((entry.ExternalAttributes >> 16) & 0xF000) == 0xA000)
                    return new(ServerDeploymentProblemCodes.PackageLayoutUnsafe, null, archiveDigest);
            }

            if (!entries.TryGetValue("manifest.json", out var manifestEntry) ||
                !entries.TryGetValue("manifest.json.sig", out var signatureEntry) ||
                manifestEntry.Length > MaximumManifestBytes || signatureEntry.Length > MaximumSignatureBytes)
                return new(ServerDeploymentProblemCodes.PackageSignatureInvalid, null, archiveDigest);

            var manifestBytes = ReadBounded(manifestEntry, MaximumManifestBytes);
            var signatureBytes = ReadBounded(signatureEntry, MaximumSignatureBytes);
            var manifest = JsonSerializer.Deserialize<ServerReleaseManifestDto>(manifestBytes, RelaxKonOSJsonOptions.Default);
            var signature = JsonSerializer.Deserialize<ServerReleaseSignatureDto>(signatureBytes, RelaxKonOSJsonOptions.Default);
            if (manifest is null || signature is null)
                return new(ServerDeploymentProblemCodes.PackageManifestInvalid, null, archiveDigest);
            if (manifest.PackageKind != expectedKind)
                return new(ServerDeploymentProblemCodes.PackageManifestInvalid, null, archiveDigest);
            if (string.IsNullOrWhiteSpace(signature.KeyId) ||
                !trustedPublicKeys.TryGetValue(signature.KeyId, out var publicKey))
                return new(ServerDeploymentProblemCodes.PackageTrustRootMissing, null, archiveDigest);

            var trust = new ServerReleaseTrustPolicy([signature.KeyId]);
            var problem = ServerReleaseValidation.VerifyManifest(
                manifestBytes, manifest, expectedRuntime, signature, publicKey, trust);
            if (problem is not null) return new(problem, null, archiveDigest);

            var signedFiles = manifest.Files.ToDictionary(item => item.Path, StringComparer.Ordinal);
            long totalLength = 0;
            foreach (var file in manifest.Files)
            {
                if (!entries.TryGetValue(file.Path, out var entry) || entry.Length != file.Length ||
                    file.Length > MaximumPayloadBytes - totalLength)
                    return new(ServerDeploymentProblemCodes.PackageManifestInvalid, null, archiveDigest);
                totalLength += file.Length;
                using var content = entry.Open();
                var digest = HashEntry(content, file.Length);
                if (!string.Equals(digest, file.Sha256, StringComparison.OrdinalIgnoreCase))
                    return new(ServerDeploymentProblemCodes.PackageDigestMismatch, null, archiveDigest);
            }

            foreach (var path in entries.Keys)
                if (path is not ("manifest.json" or "manifest.json.sig") && !signedFiles.ContainsKey(path))
                    return new(ServerDeploymentProblemCodes.PackageLayoutUnsafe, null, archiveDigest);

            return new(null, manifest, archiveDigest);
        }
        catch (Exception error) when (error is InvalidDataException or JsonException or OverflowException)
        {
            return new(ServerDeploymentProblemCodes.PackageManifestInvalid, null, null);
        }
        finally
        {
            archiveStream.Position = originalPosition;
        }
    }

    private static byte[] ReadBounded(ZipArchiveEntry entry, int maximum)
    {
        if (entry.Length > maximum) throw new InvalidDataException("Release metadata exceeds its size limit.");
        using var content = entry.Open();
        using var buffer = new MemoryStream();
        var chunk = new byte[8192];
        int count;
        while ((count = content.Read(chunk)) > 0)
        {
            if (count > maximum - buffer.Length)
                throw new InvalidDataException("Release metadata exceeds its size limit.");
            buffer.Write(chunk, 0, count);
        }
        return buffer.ToArray();
    }

    private static string HashStream(Stream stream)
    {
        using var sha = SHA256.Create();
        return Convert.ToHexString(sha.ComputeHash(stream)).ToLowerInvariant();
    }

    private static string? HashEntry(Stream stream, long expectedLength)
    {
        using var hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
        var buffer = new byte[81920];
        long received = 0;
        int count;
        while ((count = stream.Read(buffer)) > 0)
        {
            if (count > expectedLength - received) return null;
            received += count;
            hash.AppendData(buffer, 0, count);
        }
        return received == expectedLength ? Convert.ToHexString(hash.GetHashAndReset()).ToLowerInvariant() : null;
    }
}

File: Client/RelaxKonOS.Client.Android/app/src/main/java/app/relaxkonos/mobile/servercenter/ServerReleaseArchiveVerifier.kt
package app.relaxkonos.mobile.servercenter

import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.RandomAccessFile
import java.security.MessageDigest
import java.util.zip.ZipEntry
import java.util.zip.ZipException
import java.util.zip.ZipFile

/** Result of validating a complete signed release ZIP before SFTP upload. */
data class ServerReleaseArchiveVerification(
    val problemCode: String?,
    val manifest: ServerReleaseManifest?,
    val archiveSha256: String?,
) {
    val verified: Boolean get() = problemCode == null && manifest != null
}

/**
 * Verifies the archive digest, pinned manifest signature, RID, layout and every signed payload file.
 * The public key is supplied by application release configuration; a key carried inside the ZIP is
 * never considered a trust root.
 */
object ServerReleaseArchiveVerifier {
    private const val MAXIMUM_MANIFEST_BYTES = 1024 * 1024
    private const val MAXIMUM_SIGNATURE_BYTES = 16 * 1024
    private const val MAXIMUM_ENTRIES = 20_000
    private const val MAXIMUM_PAYLOAD_BYTES = 8L * 1024 * 1024 * 1024

    fun verify(
        archiveFile: File,
        expectedKind: ServerReleasePackageKind,
        expectedRuntime: ServerRuntimeIdentifier,
        trustedPublicKeys: Map<String, String>,
        expectedArchiveSha256: String? = null,
    ): ServerReleaseArchiveVerification {
        require(archiveFile.isFile) { "Release archive must be a readable file." }
        val archiveDigest = archiveFile.inputStream().use { sha256Hex(it)!! }
        if (expectedArchiveSha256 != null &&
            (!ServerDeploymentInputRules.isSha256(expectedArchiveSha256) ||
                !archiveDigest.equals(expectedArchiveSha256, ignoreCase = true))
        ) return failed(ServerDeploymentProblemCodes.PACKAGE_DIGEST_MISMATCH, archiveDigest)

        return try {
            if (!hasSafeCentralDirectory(archiveFile)) {
                return failed(ServerDeploymentProblemCodes.PACKAGE_LAYOUT_UNSAFE, archiveDigest)
            }
            ZipFile(archiveFile).use { archive ->
                val enumeration = archive.entries()
                val entries = linkedMapOf<String, ZipEntry>()
                var count = 0
                while (enumeration.hasMoreElements()) {
                    val entry = enumeration.nextElement()
                    if (++count > MAXIMUM_ENTRIES) {
                        return failed(ServerDeploymentProblemCodes.PACKAGE_LAYOUT_UNSAFE, archiveDigest)
                    }
                    val normalized = entry.name.trimEnd('/')
                    if (!ServerReleaseValidation.isSafeManifestPath(normalized)) {
                        return failed(ServerDeploymentProblemCodes.PACKAGE_LAYOUT_UNSAFE, archiveDigest)
                    }
                    if (!entry.isDirectory && entries.put(entry.name, entry) != null) {
                        return failed(ServerDeploymentProblemCodes.PACKAGE_LAYOUT_UNSAFE, archiveDigest)
                    }
                }

                val manifestEntry = entries["manifest.json"]
                    ?: return failed(ServerDeploymentProblemCodes.PACKAGE_SIGNATURE_INVALID, archiveDigest)
                val signatureEntry = entries["manifest.json.sig"]
                    ?: return failed(ServerDeploymentProblemCodes.PACKAGE_SIGNATURE_INVALID, archiveDigest)
                val manifestBytes = archive.getInputStream(manifestEntry).use {
                    readBounded(it, manifestEntry.size, MAXIMUM_MANIFEST_BYTES)
                }
                val signatureBytes = archive.getInputStream(signatureEntry).use {
                    readBounded(it, signatureEntry.size, MAXIMUM_SIGNATURE_BYTES)
                }
                val manifest = ServerDeploymentWire.readManifest(manifestBytes)
                val signature = ServerDeploymentWire.readSignature(signatureBytes)
                if (manifest.packageKind != expectedKind) {
                    return failed(ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID, archiveDigest)
                }
                val publicKey = trustedPublicKeys[signature.keyId]
                    ?: return failed(ServerDeploymentProblemCodes.PACKAGE_TRUST_ROOT_MISSING, archiveDigest)
                val signatureProblem = ServerReleaseValidation.verifyManifest(
                    manifestBytes = manifestBytes,
                    manifest = manifest,
                    expectedRuntime = expectedRuntime,
                    signature = signature,
                    publicKeyPem = publicKey,
                    trustPolicy = ServerReleaseTrustPolicy(listOf(signature.keyId)),
                )
                if (signatureProblem != null) return failed(signatureProblem, archiveDigest)

                val signedFiles = manifest.files.associateBy { it.path }
                if (signedFiles.size != manifest.files.size) {
                    return failed(ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID, archiveDigest)
                }
                var totalLength = 0L
                for (file in manifest.files) {
                    val entry = entries[file.path]
                        ?: return failed(ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID, archiveDigest)
                    if (entry.size != file.length || file.length < 0 ||
                        file.length > MAXIMUM_PAYLOAD_BYTES - totalLength
                    ) return failed(ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID, archiveDigest)
                    totalLength += file.length
                    val actual = archive.getInputStream(entry).use { sha256Hex(it, file.length) }
                    if (actual == null || !actual.equals(file.sha256, ignoreCase = true)) {
                        return failed(ServerDeploymentProblemCodes.PACKAGE_DIGEST_MISMATCH, archiveDigest)
                    }
                }
                if (entries.keys.any { it != "manifest.json" && it != "manifest.json.sig" && it !in signedFiles }) {
                    return failed(ServerDeploymentProblemCodes.PACKAGE_LAYOUT_UNSAFE, archiveDigest)
                }
                ServerReleaseArchiveVerification(null, manifest, archiveDigest)
            }
        } catch (_: ZipException) {
            failed(ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID, archiveDigest)
        } catch (_: IOException) {
            failed(ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID, archiveDigest)
        } catch (_: IllegalArgumentException) {
            failed(ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID, archiveDigest)
        } catch (_: ArithmeticException) {
            failed(ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID, archiveDigest)
        }
    }

    private fun failed(code: String, digest: String?) = ServerReleaseArchiveVerification(code, null, digest)

    private fun readBounded(input: InputStream, declaredLength: Long, maximum: Int): ByteArray {
        require(declaredLength in 0..maximum.toLong()) { "Release metadata exceeds its size limit." }
        val output = java.io.ByteArrayOutputStream(declaredLength.toInt())
        val buffer = ByteArray(8192)
        while (true) {
            val count = input.read(buffer)
            if (count < 0) break
            require(output.size() <= maximum - count) { "Release metadata exceeds its size limit." }
            output.write(buffer, 0, count)
        }
        require(output.size().toLong() == declaredLength) { "Release metadata length changed." }
        return output.toByteArray()
    }

    private fun sha256Hex(input: InputStream, expectedLength: Long? = null): String? {
        val digest = MessageDigest.getInstance("SHA-256")
        val buffer = ByteArray(81920)
        var received = 0L
        while (true) {
            val count = input.read(buffer)
            if (count < 0) break
            if (expectedLength != null && count > expectedLength - received) return null
            received += count
            digest.update(buffer, 0, count)
        }
        if (expectedLength != null && received != expectedLength) return null
        return digest.digest().joinToString("") { "%02x".format(it) }
    }

    /**
     * `ZipEntry` does not expose Unix mode bits. Inspect the central directory directly so a signed
     * entry marked as a symbolic link is rejected before upload, matching the desktop verifier.
     */
    private fun hasSafeCentralDirectory(file: File): Boolean = RandomAccessFile(file, "r").use { archive ->
        if (archive.length() < END_MINIMUM_SIZE) return false
        val searchLength = minOf(archive.length(), END_MAXIMUM_SEARCH.toLong()).toInt()
        val tail = ByteArray(searchLength)
        archive.seek(archive.length() - searchLength)
        archive.readFully(tail)
        var end = tail.size - END_MINIMUM_SIZE
        while (end >= 0 && littleInt(tail, end) != END_SIGNATURE) end--
        if (end < 0) return false
        val entries = littleShort(tail, end + 10)
        val directorySize = littleUInt(tail, end + 12)
        val directoryOffset = littleUInt(tail, end + 16)
        if (entries == 0xffff || directorySize == 0xffffffffL || directoryOffset == 0xffffffffL ||
            directoryOffset + directorySize > archive.length()
        ) return false

        archive.seek(directoryOffset)
        repeat(entries) {
            val header = ByteArray(CENTRAL_FIXED_SIZE)
            archive.readFully(header)
            if (littleInt(header, 0) != CENTRAL_SIGNATURE) return false
            val flags = littleShort(header, 8)
            if (flags and 1 != 0) return false // encrypted entries are never valid release payloads
            val nameLength = littleShort(header, 28)
            val extraLength = littleShort(header, 30)
            val commentLength = littleShort(header, 32)
            val externalAttributes = littleUInt(header, 38)
            val unixType = (externalAttributes ushr 16) and 0xF000
            if (unixType == 0xA000L) return false
            archive.seek(archive.filePointer + nameLength + extraLength + commentLength)
        }
        archive.filePointer == directoryOffset + directorySize
    }

    private fun littleShort(bytes: ByteArray, offset: Int): Int =
        (bytes[offset].toInt() and 0xff) or ((bytes[offset + 1].toInt() and 0xff) shl 8)

    private fun littleInt(bytes: ByteArray, offset: Int): Int =
        littleUInt(bytes, offset).toInt()

    private fun littleUInt(bytes: ByteArray, offset: Int): Long =
        littleShort(bytes, offset).toLong() or (littleShort(bytes, offset + 2).toLong() shl 16)

    private const val END_SIGNATURE = 0x06054b50
    private const val CENTRAL_SIGNATURE = 0x02014b50
    private const val END_MINIMUM_SIZE = 22
    private const val END_MAXIMUM_SEARCH = 65_557
    private const val CENTRAL_FIXED_SIZE = 46
}
